server/app/services/apikey/ApiKeyService.java

```diff
@@ -0,0 +1,291 @@
+package services.apikey;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import auth.ApiKeyGrants.Permission;
+import auth.CiviFormProfile;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.hash.HashCode;
+import com.google.common.hash.HashFunction;
+import com.google.common.hash.Hashing;
+import com.google.inject.Inject;
+import com.typesafe.config.Config;
+import java.nio.charset.StandardCharsets;
+import java.security.NoSuchAlgorithmException;
+import java.time.Instant;
+import java.time.format.DateTimeParseException;
+import java.util.Base64;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import javax.crypto.KeyGenerator;
+import models.ApiKey;
+import org.apache.commons.net.util.SubnetUtils;
+import play.Environment;
+import play.data.DynamicForm;
+import repository.ApiKeyRepository;
+import services.DateConverter;
+import services.program.ProgramNotFoundException;
+import services.program.ProgramService;
+
+/**
+ * Service management of the resource backed by {@link ApiKey}.
+ *
+ * <p>CiviForm uses API keys to implement HTTP basic auth
+ * (https://en.wikipedia.org/wiki/Basic_access_authentication) for authenticating API requests, as
+ * well as for authorizing requests based on the permission grants stored with the key in the
+ * database.
+ *
+ * <p>For a given key, CiviForm stores the key secret in plain text along with the key secret hashed
+ * with a secret salt value provided to the server as a configuration variable.
+ *
+ * <p>The plaintext secret is revealed to the admin user once after creation, after which it is not
+ * recoverable from CiviForm.
+ */
+public class ApiKeyService {
+
+  public static final String FORM_FIELD_NAME_KEY_NAME = "keyName";
+  public static final String FORM_FIELD_NAME_EXPIRATION = "expiration";
+  public static final String FORM_FIELD_NAME_SUBNET = "subnet";
+
+  private static final int KEY_ID_LENGTH = 128;
+  private static final int KEY_SECRET_LENGTH = 256;
+
+  private final Environment environment;
+  private final ProgramService programService;
+  private final ApiKeyRepository repository;
+  private final String secretSalt;
+  private final boolean banGlobalSubnet;
+
+  @Inject
+  public ApiKeyService(
+      ApiKeyRepository repository,
+      Environment environment,
+      Config config,
+      ProgramService programService) {
+    this.environment = checkNotNull(environment);
```

Comment:
[nit] it's nice to put these in the same order as the method params

Reply:
done